graphbolt/src/partitioned_cache_policy.cc

```diff
@@ -0,0 +1,215 @@
+  auto output_indices_ptr = output_indices.data_ptr<int64_t>();
+  torch::parallel_for(0, policies_.size(), 1, [&](int64_t begin, int64_t end) {
+    if (begin == end) return;
+    const auto tid = begin;
+    auto out_index_ptr = indices_ptr + offsets_ptr[tid];
+    begin = result_offsets[tid];
+    end = result_offsets[tid + 1];
+    const auto num_selected = end - begin;
+    auto indices_ptr = std::get<1>(results[tid]).data_ptr<int64_t>();
+    for (int64_t i = 0; i < num_selected; i++) {
+      output_indices_ptr[begin + i] = out_index_ptr[indices_ptr[i]];
+    }
+    auto selected_positions_ptr = std::get<0>(results[tid]).data_ptr<int64_t>();
+    std::transform(
+        selected_positions_ptr, selected_positions_ptr + num_selected,
+        positions.data_ptr<int64_t>() + begin,
+        [off = tid * capacity_ / policies_.size()](auto x) { return x + off; });
+    begin = result_offsets[policies_.size() + tid];
+    end = result_offsets[policies_.size() + tid + 1];
+    const auto num_missing = end - begin;
+    for (int64_t i = 0; i < num_missing; i++) {
+      output_indices_ptr[begin + i] =
+          out_index_ptr[indices_ptr[i + num_selected]];
+    }
+    std::memcpy(
+        reinterpret_cast<std::byte*>(missing_keys.data_ptr()) +
+            (begin - positions.size(0)) * missing_keys.element_size(),
+        std::get<2>(results[tid]).data_ptr(),
+        num_missing * missing_keys.element_size());
+  });
+  return std::make_tuple(positions, output_indices, missing_keys);
+}
+
+torch::Tensor PartitionedCachePolicy::Replace(torch::Tensor keys) {
+  if (policies_.size() == 1) return policies_[0]->Replace(keys);
+  torch::Tensor offsets, indices, permuted_keys;
+  std::tie(offsets, indices, permuted_keys) = Partition(keys);
+  auto output_positions =
+      torch::empty_like(keys, keys.options().dtype(torch::kInt64));
+  auto offsets_ptr = offsets.data_ptr<int64_t>();
+  auto indices_ptr = indices.data_ptr<int64_t>();
+  auto output_positions_ptr = output_positions.data_ptr<int64_t>();
+  torch::parallel_for(0, policies_.size(), 1, [&](int64_t begin, int64_t end) {
+    if (begin == end) return;
+    const auto tid = begin;
+    begin = offsets_ptr[tid];
+    end = offsets_ptr[tid + 1];
+    torch::Tensor positions =
+        policies_.at(tid)->Replace(permuted_keys.slice(0, begin, end));
+    auto positions_ptr = positions.data_ptr<int64_t>();
+    const auto off = tid * capacity_ / policies_.size();
+    for (int64_t i = 0; i < positions.size(0); i++) {
+      output_positions_ptr[indices_ptr[begin + i]] = positions_ptr[i] + off;
+    }
+  });
+  return output_positions;
+}
+
+template <typename CachePolicy>
+c10::intrusive_ptr<PartitionedCachePolicy> PartitionedCachePolicy::Create(
+    int64_t capacity, int64_t num_partitions) {
+  static_assert(std::is_base_of_v<BaseCachePolicy, CachePolicy>);
+  return c10::make_intrusive<PartitionedCachePolicy>(
+      CachePolicy(), capacity, num_partitions);
+}
```

Comment:
nit: This may cause num_partitions = 1 uses the Partitioned related code, not sure whether it will affect performance or not.

I was thinking since something like:
```
c10::intrusive_ptr<CachePolicy> PartitionedCachePolicy::Create(
    int64_t capacity, int64_t num_partitions) {
  static_assert(std::is_base_of_v<BaseCachePolicy, CachePolicy>);
  if (num_partitions == 1) return S3FifoCachePolicy;
  return c10::make_intrusive<PartitionedCachePolicy>(
      CachePolicy(), capacity, num_partitions);
}
```

Reply:
There is shortcuts now inside the partitioned query and replace functions. So there won't be any performance effect.